sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnInvokers.java

```diff
@@ -0,0 +1,631 @@
+import net.bytebuddy.implementation.bytecode.member.MethodVariableAccess;
+import net.bytebuddy.jar.asm.Label;
+import net.bytebuddy.jar.asm.MethodVisitor;
+import net.bytebuddy.jar.asm.Opcodes;
+import net.bytebuddy.matcher.ElementMatchers;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import javax.annotation.Nullable;
+
+/** Dynamically generates {@link DoFnInvoker} instances for invoking a {@link DoFn}. */
+public abstract class DoFnInvokers {
+  private DoFnInvokers() {}
+
+  public static Class<?> getDoFnClass(OldDoFn<?, ?> fn) {
+    if (fn instanceof SimpleDoFnAdapter) {
+      return ((SimpleDoFnAdapter<?, ?>) fn).fn.getClass();
+    } else {
+      return fn.getClass();
+    }
+  }
+
+  /** Create a {@link OldDoFn} that the {@link DoFn}. */
+  public static <InputT, OutputT> OldDoFn<InputT, OutputT> toOldDoFn(DoFn<InputT, OutputT> fn) {
+    DoFnSignature signature = DoFnReflector.getSignature(fn.getClass());
+    if (signature.getProcessElement().usesSingleWindow()) {
+      return new WindowDoFnAdapter<>(fn);
+    } else {
+      return new SimpleDoFnAdapter<>(fn);
+    }
+  }
+
+  /** @return the {@link DoFnInvoker} for the given {@link DoFn}. */
+  public static <InputT, OutputT> DoFnInvoker<InputT, OutputT> newByteBuddyInvoker(
+      DoFn<InputT, OutputT> fn) {
+    try {
+      @SuppressWarnings("unchecked")
+      DoFnInvoker<InputT, OutputT> invoker =
+          (DoFnInvoker<InputT, OutputT>)
+              getByteBuddyInvokerConstructor(fn.getClass()).newInstance(fn);
+      return invoker;
+    } catch (InstantiationException
+        | IllegalAccessException
+        | IllegalArgumentException
+        | InvocationTargetException
+        | SecurityException e) {
+      throw new RuntimeException("Unable to bind invoker for " + fn.getClass(), e);
+    }
+  }
+
+  private static final Map<Class<?>, Constructor<?>> BYTE_BUDDY_INVOKER_CONSTRUCTOR_CACHE =
+      new LinkedHashMap<>();
+
+  private static synchronized Constructor<?> getByteBuddyInvokerConstructor(
+      Class<? extends DoFn> fnClass) {
+    Constructor<?> constructor = BYTE_BUDDY_INVOKER_CONSTRUCTOR_CACHE.get(fnClass);
+    if (constructor != null) {
+      return constructor;
+    }
+    Class<? extends DoFnInvoker<?, ?>> wrapperClass = createWrapperClass(fnClass);
+    try {
+      constructor = wrapperClass.getConstructor(fnClass);
+    } catch (IllegalArgumentException | NoSuchMethodException | SecurityException e) {
+      throw new RuntimeException(e);
+    }
+    BYTE_BUDDY_INVOKER_CONSTRUCTOR_CACHE.put(fnClass, constructor);
+    return constructor;
+  }
+
+  private static class ContextAdapter<InputT, OutputT> extends DoFn<InputT, OutputT>.Context
```

Comment:
`/** Implement ExtraContextFactory on top of OldDoFn. */`

I believe this should go away once `DoFn` is executed directly (instead of by wrapping into an `OldDoFn`).